contracts/StablecoinConverter.sol

```diff
@@ -209,20 +215,46 @@ contract StablecoinConverter is EpochTokenLocker {
                 executedSellAmount
             );
         }
-        checkAndOverrideObjectiveValue(uint(tokenConservation[0]));
+        int objectiveValue = utility - disregardedUtility;
+        checkAndOverrideObjectiveValue(objectiveValue);
         grantRewardToSolutionSubmitter(uint(tokenConservation[0]) / 2);
         checkTokenConservation(tokenConservation);
         documentTrades(batchIndex, owners, orderIds, volumes, tokenIdsForPrice);
     }
 
-    function getCurrentObjectiveValue() public view returns(uint) {
+    function getCurrentObjectiveValue() public view returns(int) {
         if (previousSolution.batchId == getCurrentBatchId() - 1) {
             return previousSolution.objectiveValue;
         } else {
-            return 0;
+            return MIN_INT;
         }
     }
 
+    function evaluateUtility(uint128 execBuy, uint128 execSell, Order memory order) internal view returns(int) {
+        // Utility = ((execBuyAmt * order.sellAmt - execSellAmt * order.buyAmt) * price.buyToken) / order.sellAmt
+        (uint128 buyAmount, uint128 sellAmount) = getBuySellAmounts(order);
+        return ((execBuy * sellAmount - execSell * buyAmount) * currentPrices[order.buyToken]) / sellAmount;
```

Comment:
We need to watch out for overflow here. Should probably convert amounts to u256 and try to keep it to two multiplications (divide before multiplying with current prices) to make sure we don't overflow.

Reply:
Yes, I had this on my mind but wanted to "get something out there" for us to look at and evaluate beforehand. Not that I have already experimented with:

```
((execBuy * sellAmount - execSell * buyAmount)  / sellAmount) * currentPrices[order.buyToken]
```
and it returns 0 in undesired places (such as our basic trade unit test, for example). Will elaborate on this at some point... but I just wanted to point out, that it won't be as simple as rearrangement of the division. Any suggestions here?

Reply:
Ya, so that is nice, but the next equation (for disregarded utility) is going to be a bit more difficult to rearrange in a meaningful way without losing information...

```
(((sellAmt - execSell) * price[buyToken]) * buyAmt) / sellAmt;
```

Perhaps its best if I evaluate `maxUtility` with the utility formula above and simply take the difference 

```
dU = maxU - U
```

Reply:
Don't know if this helps but could we try using [fixed point numbers](https://solidity.readthedocs.io/en/latest/types.html#fixed-point-numbers)? It looks like its not fully supported yet but maybe enough for us to use? We are operating on 128bit integers we are wasting 128 bits of precision (since, if I understand correctly, all math ops are done on 256 bit integers). We could use fixed196x64 which gives us 64 bits of decimal precision and 64 extra bits to handle some overflow in the multiplications. It may provide more accurate results.

Reply:
This is a cool idea, although I found the following warning at the top of the description:

```
Fixed point numbers are not fully supported by Solidity yet. They can be declared, but cannot be assigned to or from.
```
I'll keep this open for now and experiment in a separate sandbox!